resolver/resolver.cc

```diff
@@ -948,10 +948,75 @@ class ResolveConstantsWalk {
             transformAncestor(ctx.withOwner(klass), singleton, ancst, isInclude);
         }
 
+        // TODO: Only do this check in --stripe-mode. Currently it does the check but skips reporting the error if
+        // not in --stripe-mode.
+        //
+        // TODO: Do rigorous perf analysis.
+        if (!ctx.file.data(ctx).isPackage()) {
+            const auto ambigDef = findAnyDefinitionAmbiguousWithCurrent(ctx);
+            if (ambigDef.has_value()) {
+                if (auto e = ctx.beginError(original.loc, core::errors::Resolver::AmbiguousDefinitionError)) {
+                    e.setHeader("Class definition is ambiguous", klass.show(ctx));
+                    e.addErrorLine((*ambigDef).loc(ctx), "Previously defined here.");
+                }
+            }
+        }
+
         nesting_ = nesting_->parent;
         return tree;
     }
 
+    const std::optional<core::SymbolRef> findAnyDefinitionAmbiguousWithCurrent(core::Context ctx) {
+        if (nesting_ == nullptr) {
+            // can't be ambiguous if nesting is null
+            return std::nullopt;
+        }
+
+        auto curSym = nesting_->scope;
+        auto curNesting = nesting_->parent;
```

Comment:
```suggestion
        auto &curNesting = nesting_->parent;
```
Avoid hitting the shared_ptr's counter

Reply:
Can't do this as I'm reassigning it in the loop bellow.